File: Archive_OPGEEProxyCH4Version/flashing.m
function [EF_FF] = flashing(Emissions, Activity, j, frac_control)
% The approach here to estimating tank flashing emissions is based on
% Zavala Araiza et al (2017). Flash emissions are categorized as either
% continuous or intermittent based on the maximum rate of dumping from
% separators (140 bbl/sep/day from Zavala Araiza is converted to a well basis 
% using 1.64, the average number of wells per site according to Enverus
% data). 

IsFlash = rand;

% Vector of flash emissions has 999 elements
RandomIndex = floor((rand*999)+1);

% If a tank is controlled, then the flash emission factor is zero.
if IsFlash < frac_control
    EF_FF = 0;
else
    % If throughput from the well to the separator is less than the dump
    % threshold, then the dumping is itermittent.
    if (Activity.prod_bbl(j)/Activity.wells(j)) < (140/1.64)
        % Intermittent
        IsIntDump = rand;
        
        % Dumping occurs more frequently the closer the well productivity
        % is to the dump threshold
        if IsIntDump < ((Activity.prod_bbl(j)/Activity.wells(j)) / (140/1.64))
            
            % kg CH4/well/day
            EF_FF = Emissions.HARC(RandomIndex)*(140/1.64);
        else
            EF_FF = 0;
        end
    else
        % Continuous
        
        % kg CH4/well/day
        EF_FF = Emissions.HARC(RandomIndex)*(Activity.prod_bbl(j)/Activity.wells(j));
    end
end
end

File: AnalysisCode_python/flashing.py
import numpy as np
import warnings
warnings.filterwarnings('ignore')

def flashing(Emissions, Activity, j, frac_control):
    """
    Estimate tank flashing emissions based on Zavala Araiza et al (2017).

    Parameters:
    Emissions (dict): Emissions data, including 'HARC' emissions vector.
    Activity (dict): Activity data for production, wells, etc.
    j (int): Index for the current data point.
    frac_control (float): Fraction of tanks controlled (between 0 and 1).

    Returns:
    float: Flash emissions factor for a single well (kg CH4/well/day).
    """

    IsFlash = np.random.rand()  # Random number to determine if the tank is controlled
    RandomIndex = np.ceil(np.random.rand() * 1000).astype(int) - 1

    # If the tank is controlled, the emission factor is zero
    if IsFlash < frac_control:
        EF_FF = 0
    else:
        # Threshold for continuous vs intermittent dumping (140 bbl/sep/day)
        dump_threshold = 140 / 1.64

        # Check if the throughput is less than the dump threshold
        if (Activity['prod_bbl'][j] / Activity['wells'][j]) < dump_threshold:
            IsIntDump = np.random.rand()  # Determine if intermittent dumping occurs

            # Dumping is more frequent when well productivity is closer to the dump threshold
            if IsIntDump < ((Activity['prod_bbl'][j] / Activity['wells'][j]) / dump_threshold):
                # Intermittent dumping: Calculate emission factor

                Emissions['HARC'] = Emissions['HARC'].flatten()
                EF_FF = Emissions['HARC'][RandomIndex] * dump_threshold
            else:
                EF_FF = 0
        else:
            # Continuous dumping: Calculate emission factor
            EF_FF = Emissions['HARC'][RandomIndex] * (Activity['prod_bbl'][j] / Activity['wells'][j])

    return EF_FF
